test/TestAppUtils/VisualTreeDumper.cs

```diff
@@ -0,0 +1,184 @@
+                return _filter.ShouldLogProperty(propertyInfo.Name);
+            }
+            public void VisitProperty(string propertyName, object value)
+            {
+                var v = _translator.PropertyValueToString(propertyName, value);
+                if (!_filter.IsKnownProperty(propertyName, v))
+                {
+                    AddPadding(_indent + 1);
+                    _sb.AppendLine(String.Format("<Property Name=\"{0}\" Value=\"{1}\" />", propertyName, v));
+                }
+            }
+
+            private void AddPadding(int numOfSpace)
+            {
+                _sb.Append("".PadRight(numOfSpace));
+            }
+        }
+
+        public static String DumpToXML(DependencyObject root, IPropertyValueTranslator translator, IFilter filter)
+        {
+
+            Visitor visitor = new Visitor(filter ?? new DefaultFilter(), translator ?? new DefaultPropertyValueTranslator());
+            WalkThroughTree(root, visitor);
+            return visitor.ToString();
+        }
+
+        private static void WalkThroughProperties(DependencyObject node, IVisitor visitor)
+        {
+            if (visitor.ShouldVisitPropertiesForNode(node))
+            {
+                var properties = node.GetType().GetProperties(BindingFlags.Public | BindingFlags.Instance);
+                foreach (var property in properties)
+                {
+                    if (visitor.ShouldVisitProperty(property))
+                    {
+                        Object value = null;
+
+                        try
+                        {
+                            value = property.GetValue(node, null);
+                        }
+                        catch (Exception)
+                        {
+                            value = "Exception";
+                        }
+                        visitor.VisitProperty(property.Name, value);
+                    }
+                }
+            }
+        }
+        private static void WalkThroughTree(DependencyObject node, IVisitor visitor)
+        {
+            if (visitor.ShouldVisitNode(node))
+            {
+                visitor.BeginVisitNode(node);
+
+                WalkThroughProperties(node, visitor);
+                for (int i = 0; i < VisualTreeHelper.GetChildrenCount(node); i++)
+                {
+                    WalkThroughTree(VisualTreeHelper.GetChild(node, i), visitor);
+                }
+
+                visitor.EndVisitNode(node);
+            }
+        }
+
+        public static readonly string NULL = "[NULL]";
+        public class DefaultFilter : IFilter
+        {
+            private static readonly string[] _propertyNamePostfixWhiteList = new string[] {"Brush", "Thickness"};
+            private static readonly string[] _propertyNameWhiteList = new string[] {"Background", "Foreground", "Padding", "Margin", "RenderSize", "Visibility", "Name"};
+            public virtual bool IsKnownProperty(string propertyName, string value)
```

Comment:
What's the difference between IsKnownProperty and ShouldLogProperty?

Reply:
They are the same. Renamed it to ShouldLogProperty.